feat(MCP Client Tool Node): Support Generic Auth Types

pascal-sochacki opened the pull request:
## Summary

Hello, I want to have my mcp server secured by oauth 2. Sadly, the mcp client tool node only supports static credentials. After looking into the code, I try to migrate the code to take advantage of the generic credentials auth types.

I think this should be a new version of the node. Furthermore, I need to take a look how to create a new version but also keep it existing workflow working.


<!--
Describe what the PR does and how to test.
Photos and videos are recommended.
-->

## Related Linear tickets, Github issues, and Community forum posts

I saw a Feature request here 
https://community.n8n.io/t/authorizing-mcp-node-with-oauth/130557 
https://community.n8n.io/t/mcp-client-tool-needs-custom-authentication/144666
so i'm not the only one 

And the MCP defines OAuth 2 as Auth see https://modelcontextprotocol.io/specification/draft/basic/authorization

<!--
Include links to **Linear ticket** or Github issue or Community forum post.
Important in order to close *automatically* and provide context to reviewers.
https://linear.app/n8n/issue/
-->
<!-- Use "closes #<issue-number>", "fixes #<issue-number>", or "resolves #<issue-number>" to automatically close issues when the PR is merged. -->

## Auth Types

### oAuth2Api
Should be implemented cause the mcp protocal build on top of it https://modelcontextprotocol.io/specification/2025-06-18/basic/authorization#roles

### httpCustomAuth
https://community.n8n.io/t/mcp-client-tool-needs-custom-authentication/144666
It will not be possible to have body in the auth credentials
https://docs.n8n.io/integrations/builtin/credentials/httprequest/#body

### httpHeaderAuth
used in the old version
### httpBearerAuth
used in the old version

### httpBasicAuth
### httpDigestAuth
### oAuth1Api
### httpQueryAuth


## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

pascal-sochacki commented on packages/@n8n/nodes-langchain/nodes/mcp/McpClientTool/McpClientTool.node.ts:
```diff
@@ -239,7 +231,22 @@ export class McpClientTool implements INodeType {
 		}
 
-		const { headers } = await getAuthHeaders(this, authentication);
+		if (authentication !== 'none') {
```
stolen from the http request tool node, not sure if this check is also needed here

Joffcom commented:
Hey @pascal-sochacki,

Thank you for your contribution. We appreciate the time and effort you’ve taken to submit this pull request.

Before we can proceed, please ensure the following:
	•	Tests are included for any new functionality, logic changes or bug fixes.
	•	The PR aligns with our contribution guidelines.

Regarding new nodes:
We no longer accept new nodes directly into the core codebase. Instead, we encourage contributors to follow our Community Node Submission Guide to publish nodes independently.

If your node integrates with an AI service that you own or represent, please email nodes@n8n.io and we will be happy to discuss the best approach.

About review timelines:
This PR has been added to our internal tracker as "GHC-3450". While we plan to review it, we are currently unable to provide an exact timeframe. Our goal is to begin reviews within a month, but this may change depending on team priorities. We will reach out when the review begins.

Thank you again for contributing to n8n.

pascal-sochacki commented:
@cubic-dev-ai rerun

pascal-sochacki committed:
Support Generic Auth Type in MCP Client Node

test

working creds

custom auth header

Update packages/@n8n/nodes-langchain/nodes/mcp/McpClientTool/v2/McpClientToolV2.node.ts

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

clean up

add unit test for oauth

pascal-sochacki commented:
Any Feedback?

natenho commented:
It would be very useful to have this

mydalat commented:
Thanks for working on this! OAuth 2.0 support for MCP Client Tool would be really valuable.

amitrintzler commented:
We need it

pascal-sochacki commented:
I think I will close this pull request... guess I will write a private custom node.

ericmyrem commented:
That would be a shame… perhaps publish it as a community node?

tcs-christian-ulrich commented:
pls dont close the PR. You can make it as community node anyway

elsmr commented:
Hi @pascal-sochacki, thank you for your contribution and sorry about the review delays on my end. This looks like a very valuable feature to add, and well implemented with versioning and tests.

I will do a proper review and testing tomorrow, and resolve the conflicts

pascal-sochacki commented:
Nice! There are some todo's i still need to address, like the custom auth type and auth a url fragment

Some Question: 

Do we really want to support all Auth Types of Generic Auth Type? Is there a way to exclude some options (like oauth1)?
I don't know how to tell the user that some ways you can configure a custom auth type are no supported like auth in body due to the reason how the mcp works...

elsmr committed:
Merge branch 'master' into pascal-sochacki/master

elsmr commented:
Brought the branch up-to-date with master, will continue testing and reviewing on monday

@pascal-sochacki Good point on Auth types that use the request body, let me get back to you on that

elsmr committed:
Move getNodeParameter out of util to resolve type issue

guidokok commented:
@pascal-sochacki from what I read and see in various mcp server setups, is that we do need some ability to add multiple (auth) / other headers in the mcp client request. Please find a way to make that possible. A custom request body seems to be the most logical. Maybe add some notes what is and what isn't possible.

pascal-sochacki commented:
@guidokok Can you elaborate or give an example of a mcp server which requires auth in the body of the request? 

In a perfect world, each mcp server should implement the [spec](https://modelcontextprotocol.io/specification/2025-06-18/basic/authorization) correctly....

guidokok commented:
@pascal-sochacki of coarse!

See Twelvedata! 

They have hundreds of endpoints and decided to use OpenAI to find out what endpoints to serve to a mcp client. 

Therefore you need there api key as well as that of OpenAI. 
Read here; https://github.com/twelvedata/mcp

pascal-sochacki commented:
I only see Headers as Auth in the twelvedata readme. So this is fine and also already possible with this Pull request

I also did write a test for this https://github.com/n8n-io/n8n/pull/17880/files#diff-ee99fbf534c49c273ee53ea072c9ebeb73bc5fb09e9e159d5feb1e64f218b977R85 (only one header but it should work with n headers)

pascal-sochacki commented:
Yes, my pull request would add all auth types listed here: https://docs.n8n.io/integrations/builtin/credentials/httprequest/#supported-authentication-methods

Custom auth is a part of this. Some Auth Types still need to be discussed. Like how to handle https://docs.n8n.io/integrations/builtin/credentials/httprequest/#body custom auth with body

elsmr reviewed (commented):
Sorry for the delays, I discussed internally and I don't think we can merge this PR in its current state. The OAuth2 implementation needs some work and will require core changes. I will prioritize this internally.

yamedoff commented:
+1 Most MCPs are adopting oauth 2.1

elsmr commented:
Hey all, we’ll be working on OAuth2 support in the MCP Client next week. I can’t share an exact release timeline, but it shouldn’t take too long. Unfortunately, we can’t move forward with this PR due to limitations around OAuth2 and token refresh.

@pascal-sochacki thank you for the time and effort you’ve put into this, we really appreciate it.

cloudsolutiongmbh commented:
+1, this is really taking too long now....

emreyeter commented:
> Hey @emreyeter, what do you mean exactly by custom auth?

basically, i want to add more than one header to mcp client tool. bearer,header options are not fit for me. i need to pass multiple headers for each mcp clients (including token)
